mclf/curves/smooth_projective_curves.py

```diff
@@ -1265,3 +1271,69 @@ def e_f_of_valuation(v):
     if hasattr(v, "_approximation"):
         v = v._approximation
     return (v.E(), v.F())
+
+
+def extension_of_finite_field(K, n):
+    r""" Return a field extension of this finite field of degree n.
+
+    INPUT:
+
+    - ``K`` -- a finite field
+    - ``n`` -- a positive integer
+
+    OUTPUT: a field extension of `K` of degree `n`.
+
+    This function is useful if `K` is constructed as an explicit extension
+    `K = K_0[x]/(f)`; then ``K.extension(n)`` is not implemented.
+
+    """
+    assert K.is_field()
+    assert K.is_finite()
+    q = K.order()
+    R = PolynomialRing(K, 'z')
+    z = R.gen()
+    # we look for a small number e dividing q^n-1 but not q-1
+    e = min([d for d in (q**n-1).divisors() if not d.divides(q-1)])
+    F = (z**e-1).factor()
+    f = [g for g, e in F if g.degree() == n][0]
+    # this is very inefficient!
+    return K.extension(f, 'z'+str(e))
+
+
+def make_finite_field(k):
+    r""" Return the finite field isomorphic to this field.
+
+    INPUT:
+
+    - ``k`` -- a finite field
+
+    OUTPUT: a pair `(k_1,\phi)` where `k_1` is a 'true' finite field
+    and `\phi` is an isomorphism from `k` to `k_1`.
+
+    This function is useful when `k` is constructed as a tower of extensions
+    with a finite field as a base field.
+
+    """
+    from sage.categories.finite_fields import FiniteFields
+
+    assert k.is_field()
+    assert k.is_finite()
+    if k in FiniteFields:
+        return k, k.hom(k.gen(), k)
+    else:
+        k0 = k.base_field()
+        G = k.modulus()
+        assert G.parent().base_ring() is k0
+        k0_new, phi0 = make_finite_field(k0)
+        G_new = G.map_coefficients(phi0, k0_new)
+        k_new = k0_new.extension(G_new.degree())
+        alpha = G_new.roots(k_new)[0][0]
+        Pk0 = k.cover_ring()
```

Comment:
That's an interesting hack to work around https://trac.sagemath.org/ticket/26105 if I understand this code correctly.